src/librustc/traits/error_reporting.rs

```diff
@@ -825,6 +820,24 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         }
         err
     }
+
+    fn report_closure_arg_mismatch(&self,
+                           span: Span,
+                           found_span: Option<Span>,
+                           expected_ref: ty::PolyTraitRef<'tcx>,
+                           found: ty::PolyTraitRef<'tcx>)
+        -> DiagnosticBuilder<'tcx>
+    {
+        let mut err = struct_span_err!(self.tcx.sess, span, E0623,
+            "type mismatch in closure arguments");
+        if let Some(sp) = found_span {
+            err.span_label(span, format!("expected closure that takes a `{}`", found));
```

Comment:
we should be able to extract the types of the arguments by examining the **substs** of the `expected_ref` and `found` trait-refs. Something like:

```
let expected_args = expected_ref.map_bound_ref(|trait_ref| trait_ref.substs.type_at(0));
```

But it might need to be `type_at(1)`. That will give you a `Binder<Ty<'tcx>>`, which you should be able to use with `{}` I think.